apiclient/src/version.rs

```diff
@@ -0,0 +1,171 @@
+// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::{
+    collections::HashSet,
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+use http::{HeaderMap, StatusCode};
+use phnxtypes::{
+    messages::{
+        client_qs::{ClientToQsMessageTbs, VersionError},
+        ApiVersion,
+    },
+    ACCEPTED_API_VERSIONS_HEADER,
+};
+use tracing::error;
+
+/// Api versions that were negotiated with the server.
+///
+/// The default values are the current API versions of the corresponding messages.
+pub(crate) struct NegotiatedApiVersions {
+    qs_api_version: AtomicU64,
+}
+
+impl NegotiatedApiVersions {
+    pub(crate) fn new() -> Self {
+        Self {
+            qs_api_version: AtomicU64::new(ClientToQsMessageTbs::CURRENT_API_VERSION.value()),
+        }
+    }
+
+    pub(crate) fn set_qs_api_version(&self, version: ApiVersion) {
+        self.qs_api_version
+            .store(version.value(), Ordering::Relaxed);
+    }
+
+    pub(crate) fn qs_api_version(&self) -> ApiVersion {
+        let version = self.qs_api_version.load(Ordering::Relaxed);
+        ApiVersion::new(version).expect("logic error")
+    }
+}
+
+/// Returns `Some` if the server supports a different API version, otherwise None.
+///
+/// If there is no API version supported by this client which is accepted by the server, the
+/// returned result is an error.
+pub(crate) fn api_version_negotiation(
+    response: &reqwest::Response,
+    current_version: ApiVersion,
+    supported_versions: &[ApiVersion],
+) -> Option<Result<ApiVersion, VersionError>> {
```

Comment:
To make `prepare_and_send_qs_message` more readable, I'd propose pulling the initial `if` check out. I find the use of `Option` to communicate whether the initial version was acceptable in the first place a little confusing.